### ISSUE TITLE
underlying cache implementation use Swift Dictionary rather than NSCache

### ISSUE THREAD

**Issue by wukongzhikong** [Alamofire/AlamofireImage#161]
Thank you for creating such a elegant image cache library, but these days when i read the source code, i found that class AutoPurgingImageCache use a Swift Dictionary as underlying cache store rather than a NSCache instance, so it would be great if you can explain the intension.
look forward to your reply!

**Comment by cnoon** [comment]
Great question @wukongzhikong! 

The main reason is that we do very accurate tracking on the memory capacity to make sure the cache purges when the memory limits are exceeded. This allows you to have very fine grained control over the caching behavior as the client. If we used an `NSCache` to store all the info, we wouldn't be able to monitor all the changes to the memory capacity and thus it would defeat the purpose of being able to explicitly control the caching behavior.

At Nike, we've built an app on top of the AFI caching system that uses multiple image caches. We have different limits set on them because they cache different types of images. If we used two `NSCache` instances in this case, we wouldn't have any control over the purging behavior.

We also have prioritized image purging based on the last accessed date. This is nice because it will always purge the images that have not been accessed recently. This helps ensure you always have the latest used images in the cache because you're most likely to hit them again.

All these things combined make using a plain ole dictionary the best data structure for the job.

Cheers. 🍻
